### ISSUE TITLE
the transport you get from a process endpoint is not a producer or consumer

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7436]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#7436|
|Type|enhancement|
|Created|2014-05-29 00:15:55Z|
|Branch|https://github.com/twisted/twisted/tree/process-producer-7436|

While implementations of `IPullProducer` and `IConsumer` are not, I guess, technically _required_ of an `IStreamClientEndpoint` - it's sure hard to do something useful with the transport if they don't.

Additionally, the underlying process transport totally does this already, so we just need to proxy a few more methods.

<details><summary>Searchable metadata</summary>

```
trac-id__7436 7436
type__enhancement enhancement
reporter__glyph glyph
priority__normal normal
milestone__ 
branch__branches_process_producer_7436 branches/process-producer-7436
branch_author__glyph glyph
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1401322555000000 1401322555000000
changetime__1403821869243018 1403821869243018
version__None None
owner__glyph glyph

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

Just trying to be intentional about classifying things more now: It's an enhancement because this is just new functionality I guess.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

(In [42751]) Branching to process-producer-7436.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban set owner to @glyph|
|-|-|

I agree that is hard to use a process transport which does not implement consumer/producer.

The interface update is great!

I had the same problem while trying to implement / emulate an in memory spawned process for SSH exec command in conch... but this
was with the old API... not with endpoints.

-------

```
# verifyClass(interfaces.IConsumer, MemoryProcessTransport)
```
Why do we need this commented line?

----------

Does this branch has a news file ?

----------------

For test_extraneousAttributes ... I think that test is ok, but it
rather test that it is a proxyForInterface... maybe this test
should check that is a proxy ForInterface and have a separate test which checks proxyForInterface behaviour.
Otherwise we will have a similar test as this one for each
class using proxyForInterface

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst commented|
|-|-|

What was the next step? "Address and resubmit" or "address and merge"?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph commented|
|-|-|

Replying to [itamar](#note_5):
> What was the next step? "Address and resubmit" or "address and merge"?

I'm going to interpret this as a "address and merge" because the feedback appears thorough enough to have considered the whole branch, and the things to change are minor enough (NEWS file, minor testing quibbles) that I would have made them an "address and merge" myself.  Also adi might well have been unable to comment all week due to our horrible spam problems :-\.

In the future though please do include explicit guidance for what to do next.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph commented|
|-|-|

Replying to [adiroiban](#note_4):
> I agree that is hard to use a process transport which does not implement consumer/producer.
> 
> The interface update is great!

Thanks!

> ```
> # verifyClass(interfaces.IConsumer, MemoryProcessTransport)
> ```
> Why do we need this commented line?

Good catch, we reinstated this and the other `verifyClass` check.

> Does this branch has a news file ?

It does now.

> For test_extraneousAttributes ... I think that test is ok, but it
> rather test that it is a proxyForInterface... maybe this test
> should check that is a proxy ForInterface and have a separate test which checks proxyForInterface behaviour.
> Otherwise we will have a similar test as this one for each
> class using proxyForInterface

The problem is that the property we want to check for - the reason we're using proxyForInterface - is to filter the extraneous attributes, so it makes sense to have a test which checks that.  I think that perhaps proxyForInterface should have some testing support for this, but given that the test only has one line of code in it, I think this is fine to live with for now.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph set status to `closed`|
|-|-|

(In [42817]) Merge process-producer-7436

Author: glyph, dreid
Reviewer: adiroiban
Fixes: [#7436](https://github.com/twisted/twisted/issues/7436)

Allow much more of the code within Twisted to use ProcessEndpoint by adding
IPushProducer and IConsumer interfaces to its resulting transport.
